### DISCUSSION TITLE
Electrostatic surface potential

### DISCUSSION THREAD

**Discussion by dhrblav** [molstar/molstar#417] [Q&A] [answered]
Hi,
I would like to visualize electrostatic surface potential in Mol* using DX files as generated by APBS. The https://www.poissonboltzmann.org website provides a nice interface to compute the electrostatic potential data and to visualize it in a 3Dmol web page (see eg. https://server.poissonboltzmann.org/viz/3dmol?jobid=2osettv94k&pqr=2osettv94k&date=2022-04-03).
![image](https://user-images.githubusercontent.com/93319094/161801695-9e0bd4a1-04f3-4aad-af74-0b33c6e210e3.png)

Ultimately I would like to do this programmatically, but is it already possible in the Mol* Viewer? I can load DX files in the viewer, but I can't figure out how to colorize the surface according to the electrostatic potential.
Am I missing something, or does this require further development?

Thank you.

**Comment by dsehnal** [comment, marked as answer]
Currently, you would have to create a custom theme that does this coloring. Maybe we can collaborate on this a little and add first class support for this?

**Reply by dhrblav** [reply]
Thank you David. I will investigate how this could be implemented and will keep you posted.
André

**Comment by dhrblav** [comment]
Back to this topic.
As far as I understand the problem, in order to create an appropriate custom color theme for electrostatic potential, we will need to retrieve potential values from the volume grid (DX file or whatever) at positions close to the protein molecular surface.
Questions arise:
-Is the molecular surface (AccessibleSurfaceArea?) systematically computed when a structure is loaded? How can it be accessed? If it is not pre-computed how to create it?
-Once we have the molecular surface are there utility functions in Mol* to retrieve potential grid points close to the surface? GridLookup3D?

**Reply by dsehnal** [reply]
There are two ways to pass the data and sample them:
- You can pass a whole grid (packed to a texture) to a color theme and the sampling will be done automatically: https://github.com/molstar/molstar/blob/master/src/mol-theme/color.ts#L73 
  - @arose any tips on what we would a good way to convert the source volume into the texture?
- You can use a theme with ``'vertex'`` granularity ([example](https://github.com/molstar/molstar/blob/master/src/examples/basic-wrapper/custom-theme.ts#L19)) and sample the grid from the position (finding the cell from a position is very easy and requires just 3 floor divisions)

To pass the raw volume data to a theme a [value ref param](https://github.com/molstar/molstar/blob/master/src/mol-util/param-definition.ts#L291) can be used.

I am thinking the whole thing could be implemented in a way similar to how the [LoadTrajectory action](https://github.com/molstar/molstar/blob/master/src/mol-plugin-state/actions/structure.ts#L318) works.

**Reply by dhrblav** [reply]
Using a theme with 'vertex' granularity does the job very nicely (the basic-wrapper custom theme example is very helpful, thanks).
Now, to pass the raw volume data to a theme, you mention valueRef params. Any hint or example on how to do this in concrete code?

**Reply by dsehnal** [reply]
Pseudo-ish code...

in params for the theme, use the ``DataRef`` param definition:

```ts
const DXThemeParams = {
  potential: PD.DataRef<Volume>()
  ...
};
```

Load and parse the DX volume:

```ts
const data = download(url);
const parsed = apply DX parse transform
```

Pass the ref to the color theme (same as you would update theme in code for any other case:

```
plugin.build().to(visual).update(... colorParams: { potential: { ref: parsed.selector.ref }  }  ...);
```

Now the value passed to the theme will automatically be resolved to the Volume data.

**Reply by dhrblav** [reply]
Excellent! Thank you

**Comment by mschroering** [comment]
Is there a full example of this somewhere?  I'm trying to do something similar and am pretty new to molstar.
